versioned/transfer/src/main/java/org/projectnessie/versioned/transfer/BoundedList.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2022 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.transfer;
+
+import com.google.common.base.Preconditions;
+import java.util.function.Consumer;
+
+/**
+ * Keeps track of the last (up to {@code capacity}) elements added to this collection.
+ *
+ * <p>Operations on this collection type:
+ *
+ * <ul>
+ *   <li>removal of the least recently added element
+ *   <li>retrieving the least recently added element
+ *   <li>consuming present elements, starting with the most or least recent one
+ * </ul>
+ */
+final class BoundedList<T> {
```

Comment:
Could we use the standard `ArrayDeque` instead?

Reply:
It's a bit tricky what the optimizing implementation does here. `ArrayDeque` is unbounded and some specific operations in `BoundedList` would need to be ported - doubt it would make the code cleaner.

Reply:
fair enough